File: Level 1/크레인 인형뽑기 게임/solution.js
function solution(board, moves) {
  const lanes = Array.from({ length: board[0].length }, () => []);

  [...board.reverse()].forEach((row) => {
    row.forEach((doll, index) => {
      if (doll) {
        lanes[index].push(doll);
      }
    });
  });

  const bucket = [];
  let answer = 0;

  moves.forEach((move) => {
    const doll = lanes[move - 1].pop();

    if (!doll) {
      return;
    }

    if (bucket.at(-1) === doll) {
      bucket.pop();
      answer += 2;
    } else {
      bucket.push(doll);
    }
  });

  return answer;
}

File: Level 1/크레인 인형뽑기 게임/solution.py
def solution(board, moves):
    lanes = [[] for _ in range(len(board[0]))]

    for row in board[::-1]:
        for index, doll in enumerate(row):
            if doll:
                lanes[index].append(doll)

    bucket = []
    answer = 0

    for move in moves:
        lane = lanes[move - 1]

        if not lane:
            continue

        doll = lane.pop()

        if bucket and bucket[-1] == doll:
            bucket.pop()
            answer += 2
        else:
            bucket.append(doll)

    return answer
